xarray/backends/pseudonetcdf_.py

```diff
@@ -0,0 +1,121 @@
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import functools
+
+import numpy as np
+
+from .. import Variable
+from ..core.pycompat import OrderedDict
+from ..core.utils import (FrozenOrderedDict, Frozen)
+from ..core import indexing
+
+from .common import AbstractDataStore, DataStorePickleMixin, BackendArray
+
+
+class PncArrayWrapper(BackendArray):
+
+    def __init__(self, variable_name, datastore):
+        self.datastore = datastore
+        self.variable_name = variable_name
+        array = self.get_array()
+        self.shape = array.shape
+        self.dtype = np.dtype(array.dtype)
+
+    def get_array(self):
+        self.datastore.assert_open()
+        return self.datastore.ds.variables[self.variable_name]
+
+    def __getitem__(self, key):
+        key, np_inds = indexing.decompose_indexer(
+            key, self.shape, indexing.IndexingSupport.OUTER_1VECTOR)
+
+        with self.datastore.ensure_open(autoclose=True):
+            array = self.get_array()[key.tuple]  # index backend array
+
+        if len(np_inds.tuple) > 0:
+            # index the loaded np.ndarray
+            array = indexing.NumpyIndexingAdapter(array)[np_inds]
+        return array
+
+
+_genericncf = ('Dataset', 'netcdf', 'ncf', 'nc')
+
+
+class _notnetcdf:
+    def __eq__(self, lhs):
+        return lhs not in _genericncf
+
+
+class PseudoNetCDFDataStore(AbstractDataStore, DataStorePickleMixin):
+    """Store for accessing datasets via PseudoNetCDF
+    """
+    @classmethod
+    def open(cls, filename, format=None, writer=None,
+             autoclose=False, **format_kwds):
+        from PseudoNetCDF._getreader import getreader, getreaderdict
```

Comment:
Importing from private APIs (with the underscore prefix) makes me nervous. In general we try to avoid doing this because private APIs can change at any time, though of course that isn't true here because you're the PNC maintainer. I'm also nervous about including constants like `_genericncf = ('Dataset', 'netcdf', 'ncf', 'nc')` in xarray because that feels like logic that should live in libraries like PNC instead.

So maybe we should stick with the standard open method that you were using earlier. In particular, if we change the default to `decode_cf=False` for PNC in `open_dataset()` then I don't think it's essential to check the format.

Reply:
I think it will be confusing for users to have an indirect use of `netCDF4` via `PseudoNetCDF` that responds differently than the typical backend. This code is designed to address that xarray-specific issue. Typical behavior in PNC is to mask_and_scale via netCDF4's default functionality, so with `decode_cf` a netCDF file would still be decoded. This is confusing to me, let alone a user. For this reason, I think it is right to exclude formats that already have specifically optimized backends.

I agree completely about private API. The standard pncopen, however, does not currently support the ability to enable a subset of readers. I will add this a feature request to PseudoNetCDF, and make a subsequent update there. As soon as I update pncopen, I would make the update here.

The `_genericncf` is not logic that should live in PNC. It is logic designed specifically to address an xarray issue. In this case, `xarray` has a feature that should pre-empt PNC typical capabilities.

The alternative is to update PNC, update the conda installation, and update this code again.

Reply:
Are you thinking of someone explicitly calling `xarray.decode_cf()` on a dataset loaded from PNC? I agree that that would give unexpected results, but I don't see that being done very often.

Another way to make all of this safe (and which wouldn't require disabling `mask_and_scale` in `open_dataset()`) would be to explicitly drop `_FillValue`, `missing_value`, `add_offset` and `scale_factor` attributes from variables loaded with PNC, moving them to `Variable.encoding` instead.

Reply:
Imagine that you have a netcdf file. If you open it with PNC, it uses netCDF4 as the library to open it. In xarray, however, you might get a different processing. If you write it out, it might have different properties. This seems undesirable at best. That is why I prefer removing the netCDF functionality.

In PNC, the goal of using netCDF4 is to make sure all pnc functionality is available for any type of file. In PNC, some plotting and processing tools are available (similar to xarray). I think plotting and processing in xarray is often better than PNC. The goal of adding PNC to xarray is to bring xarray formats that PNC can read. Since xarray already has mechanisms for supporting netCDF, I think bringing this functionality is unnecessary.

Reply:
My general feeling is that we should make it easy to do the right thing, but not put up artificial barriers to stop expert users. I think this is broadly consistent with most Python projects, including the Python language itself. Even if it's strange, somebody might have reasons for trying PNC/netCDF4 together, and as long as it will not give wrong results there is no reason to disable it.

That said, my real objection here are all the private imports from PNC, and the four different aliases we use to check for netCDF files. If you want to put a public API version of this in PNC before merging this PR, that would be OK with me.

Reply:
Okay. I removed the type checking and private imports. I updated the testing. 

I get one failed that I cannot identify as related to PNC: `test_open_mfdataset_manyfiles` files with:

```
E           OSError: [Errno 24] Too many open files: '/var/folders/g2/hwlpd21j4vl8hvldb4tyrz400000gn/T/tmp0bteny75'
```

I'm going to run it through the build to see if it is just my setup.

Reply:
It didn't fail on the automated test, so all passing [x] and no private imports [x].